### PR TITLE
`DataLoader`s 5: add support for external binary `DataLoader`s (PATH)

### PR DESCRIPTION
**Pull request by teh-cmc**
Turn any executable in your `PATH` into a `DataLoader` by logging data to stdout. 

Includes examples for all supported languages.

image

Checks:
- [x] `rerun-loader-rust-file` just prints help message
- [x] `rerun-loader-rust-file examples/rust/external_data_loader/src/main.rs` successfully does nothing
- [x] `rerun-loader-rust-file examples/rust/external_data_loader/src/main.rs | rerun -` works
- [x] `rerun-loader-python-file` just prints help message
- [x] `rerun-loader-python-file examples/python/external_data_loader/main.py` successfully does nothing
- [x] `rerun-loader-python-file examples/python/external_data_loader/main.py | rerun -` works
- [x] `rerun-loader-cpp-file` just prints help message
- [x] `rerun-loader-cpp-file examples/cpp/external_data_loader/main.cpp` successfully does nothing
- [x] `rerun-loader-cpp-file examples/cpp/external_data_loader/main.cpp | rerun -` works
- [x] `cargo r -p rerun-cli --no-default-features --features native_viewer -- examples/assets`
- [x] Native: `File > Open > examples/assets/`
- [x] Native: `Drag-n-drop > examples/assets/`

---

Part of a series of PRs to make it possible to load _any_ file from the local filesystem, by any means, on web and native:
- #4516
- #4517 
- #4518 
- #4519 
- #4520 
- #4521 
- TODO: register custom loaders
- TODO: high level docs and guides for everything related to loading files

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Full build: [app.rerun.io](https://app.rerun.io/pr/4516/index.html)
  * Partial build: [app.rerun.io](https://app.rerun.io/pr/4516/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json) - Useful for quick testing when changes do not affect examples in any way
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4516)
- [Docs preview](https://rerun.io/preview/f83cf44a7283e89a454d3d299b81a75bea2e4a50/docs) 
- [Examples preview](https://rerun.io/preview/f83cf44a7283e89a454d3d299b81a75bea2e4a50/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

**Comment by teh-cmc** [comment]
I forgot to update the `DataLoader` docstring to link the new loader, will do so in the docs/howto guide PR.

File: crates/re_data_source/src/data_loader/loader_external.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
+use std::io::Read;
+
+use once_cell::sync::Lazy;
+
+/// To register a new external data loader, simply add an executable in your $PATH whose name
+/// starts with this prefix.
+pub const EXTERNAL_DATA_LOADER_PREFIX: &str = "rerun-loader";
```
Comment (emilk) [inline crates/re_data_source/src/data_loader/loader_external.rs:7]:
```suggestion
pub const EXTERNAL_DATA_LOADER_PREFIX: &str = "rerun-loader-";
```

🤷

File: crates/re_data_source/src/data_loader/loader_external.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        re_tracing::profile_function!(filepath.display().to_string());
+
+        for exe in EXTERNAL_LOADER_PATHS.iter() {
+            let store_id = store_id.clone();
+            let filepath = filepath.clone();
+            let tx = tx.clone();
+
+            // NOTE: spawn is fine, the entire loader is native-only.
+            rayon::spawn(move || {
+                let child = Command::new(exe)
```
Comment (emilk) [inline crates/re_data_source/src/data_loader/loader_external.rs:88]: add a profile scope for the spawned task

File: examples/rust/external_data_loader/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+// It is up to you whether you make use of that shared recording ID or not.
+// If you use it, the data will end up in the same recording as all other plugins interested in
+// that file, otherwise you can just create a dedicated recording for it. Or both.
+struct Args {
+    filepath: std::path::PathBuf,
+    recording_id: Option<String>,
+}
+
+impl Args {
+    fn from_env() -> Result<Self, pico_args::Error> {
```
Comment (emilk) [inline examples/rust/external_data_loader/src/main.rs:43]: @jprochazk just started using `argh` for our build scripts (compiles much faster than `clap`); you might want to use that here too for consistency (and so we don't have three different argument parsers)

File: crates/re_data_source/src/load_file.rs [hunk 4]
```diff
@@ -137,10 +137,17 @@ pub(crate) fn load(
     is_dir: bool,
     contents: Option<std::borrow::Cow<'_, [u8]>>,
 ) -> Result<std::sync::mpsc::Receiver<LoadedData>, DataLoaderError> {
+    #[cfg(target_arch = "wasm32")]
+    let no_external_loaders = true;
+    #[cfg(not(target_arch = "wasm32"))]
+    let no_external_loaders = crate::data_loader::EXTERNAL_LOADER_PATHS.is_empty();
```
Comment (emilk) [inline crates/re_data_source/src/load_file.rs:143]: don't put negations in names

```suggestion
    #[cfg(target_arch = "wasm32")]
    let has_external_loaders = false;
    #[cfg(not(target_arch = "wasm32"))]
    let has_external_loaders = !crate::data_loader::EXTERNAL_LOADER_PATHS.is_empty();
```

File: crates/re_data_source/src/data_loader/loader_external.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
+use std::io::Read;
```
Comment (emilk) [inline crates/re_data_source/src/data_loader/loader_external.rs:1]: You should document somewhere what we expect from a loader, eg in a `//!`-level docstring

In particular:
* what arguments to expect (`--recording-id`)
* what exit code to use for unsupported file types
* what and where to output (.rrd data to `stdout`)

This should also be added as a how-to guide on rerun.io/docs!

**Review by emilk — APPROVED** [review]

File: crates/re_data_source/src/data_loader/loader_external.rs (hunk 5 above)
Comment (teh-cmc) [reply crates/re_data_source/src/data_loader/loader_external.rs:1]: I have an upcoming PR that adds a web guide, I'll add a link to that guide in the doc string then

**Pull request merged**